pkg/controller/controllerutil/controllerutil.go

```diff
@@ -121,57 +121,54 @@ const ( // They should complete the sentence "Deployment default/foo has been ..
 //
 // It returns the executed operation and an error.
 func CreateOrUpdate(ctx context.Context, c client.Client, obj runtime.Object, f MutateFn) (OperationResult, error) {
-	// op is the operation we are going to attempt
-	op := OperationResultNone
-
-	// get the existing object meta
-	metaObj, ok := obj.(v1.Object)
-	if !ok {
-		return OperationResultNone, fmt.Errorf("%T does not implement metav1.Object interface", obj)
+	key, err := client.ObjectKeyFromObject(obj)
+	if err != nil {
+		return OperationResultNone, err
 	}
 
-	// retrieve the existing object
-	key := client.ObjectKey{
-		Name:      metaObj.GetName(),
-		Namespace: metaObj.GetNamespace(),
+	if err := c.Get(ctx, key, obj); err != nil {
+		if errors.IsNotFound(err) {
+			if err := safeMutate(key, obj, f); err != nil {
+				return OperationResultNone, err
+			}
+
+			if err := c.Create(ctx, obj); err != nil {
+				return OperationResultNone, err
+			}
+			return OperationResultCreated, nil
+		}
+		return OperationResultNone, err
 	}
-	err := c.Get(ctx, key, obj)
 
-	// reconcile the existing object
 	existing := obj.DeepCopyObject()
-	existingObjMeta := existing.(v1.Object)
-	existingObjMeta.SetName(metaObj.GetName())
-	existingObjMeta.SetNamespace(metaObj.GetNamespace())
-
-	if e := f(obj); e != nil {
-		return OperationResultNone, e
+	if err := safeMutate(key, obj, f); err != nil {
+		return OperationResultNone, err
 	}
 
-	if metaObj.GetName() != existingObjMeta.GetName() {
-		return OperationResultNone, fmt.Errorf("ReconcileFn cannot mutate objects name")
+	if reflect.DeepEqual(existing, obj) {
+		return OperationResultNone, nil
 	}
 
-	if metaObj.GetNamespace() != existingObjMeta.GetNamespace() {
-		return OperationResultNone, fmt.Errorf("ReconcileFn cannot mutate objects namespace")
+	if err := c.Update(ctx, obj); err != nil {
+		return OperationResultNone, err
 	}
+	return OperationResultUpdated, nil
+}
 
-	if errors.IsNotFound(err) {
-		err = c.Create(ctx, obj)
-		op = OperationResultCreated
-	} else if err == nil {
-		if reflect.DeepEqual(existing, obj) {
-			return OperationResultNone, nil
-		}
-		err = c.Update(ctx, obj)
-		op = OperationResultUpdated
-	} else {
-		return OperationResultNone, err
+func safeMutate(key client.ObjectKey, obj runtime.Object, f MutateFn) error {
+	if err := f(obj); err != nil {
+		return err
 	}
 
+	newKey, err := client.ObjectKeyFromObject(obj)
 	if err != nil {
-		op = OperationResultNone
+		return err
+	}
+
+	if key != newKey {
+		return fmt.Errorf("MutateFn cannot mutate object key")
```

Comment:
Error message will be more useful if we point out which fields are not mutable.
e.g. `MutateFn cannot mutate namespace and (or) object name.`